test/python/test_fake_signer.py

```diff
@@ -89,70 +86,66 @@ def test_fake_partner_credentials_signed(self, backend, operation):
 
     # CHECK THAT A TRANSACTION INFORMATION NOT SIGNED BY THE PARTNER KEY IS REFUSED
 
-    @pytest.mark.parametrize("operation", ["SWAP", "FUND", "SELL"])
+    @pytest.mark.parametrize("operation", [SubCommand.SWAP, SubCommand.FUND, SubCommand.SELL])
     def test_fake_transaction_infos(self, backend, operation):
-        ex = ExchangeClient(backend, Rate.FIXED, SUB_COMMAND[operation])
-        partner = SigningAuthority(curve=ex.partner_curve, name="partner")
-        partner_fake = SigningAuthority(curve=ex.partner_curve, name="partner_fake")
+        ex = ExchangeClient(backend, Rate.FIXED, operation)
+        partner = SigningAuthority(curve=get_partner_curve(operation), name="partner")
+        partner_fake = SigningAuthority(curve=get_partner_curve(operation), name="partner_fake")
 
-        ex.init_transaction()
+        transaction_id = ex.init_transaction().data
         ex.set_partner_key(partner.credentials)
         ex.check_partner_key(LEDGER_SIGNER.sign(partner.credentials))
-        ex.process_transaction(TX_INFOS[operation], FEES)
+        tx = craft_tx(operation, TX_INFOS[operation], transaction_id)
+        ex.process_transaction(tx, FEES)
```

Comment:
This is the way it was done before, and it was very hard to introduce errors / valid edge cases
In your code snippet, it will be very hard to send a wrong subcommand for the expected transaction for example

However! I do agree that there is a LOT of repetitiveness, but I plan solve it by introducing helpers that perform valid tx flows, (and calling them everywhere we don't do an error), much like the `ExchangeTestRunner`

In the end I plan to have something like this :
```
def test_error_at_step_4(exchange_client, helpersomething):
    helpersomething.run_until_step_4()
    exchange_client.step4(false data)
```

Reply:
So yes there is a lot of redundancy and it's an issue, but putting it in the client is IMO not the good way to abstract it